File: app/lib/src/main/java/com/example/binarySearch/FirstAndLastPosition6.java
package com.example.binarySearch;

import java.util.Arrays;

public class FirstAndLastPosition6 {
    public static void main(String[] args) {
        int[] arr = {5, 7, 7, 8, 8, 10};
        int target = 8;
        int[] res=searchRange(arr,target);
        System.out.println(Arrays.toString(res));;
    }

    public static int[] searchRange(int[] nums, int target) {
        int[] ans = {-1, -1};
        // check for first occurrence if target first
        int start = search(nums, target, true);
        int end = search(nums, target, false);
        ans[0] = start;
        ans[1] = end;
        return ans;
    }

    // this function just returns the index value of target
    static int search(int[] nums, int target, boolean findStartIndex) {
        int ans = -1;
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (target < nums[mid]) {
                end = mid - 1;
            } else if (target > nums[mid]) {
                start = mid + 1;
            } else {
                ans = mid;
                if (findStartIndex) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
        }
        return ans;
    }
}

File: app/lib2/src/main/java/com/example/lib2/kunal/dsa/binarySearch3/FirstAndLastPosition6.kt
package com.example.lib2.kunal.dsa.binarySearch3

fun main() {
    var array = arrayOf(5, 7, 7, 8, 8, 10)
    var target = 8
    println(searchRange(array, target).contentToString())
}

fun searchRange(nums: Array<Int>, target: Int): Array<Int> {
    val ans = arrayOf(-1, -1)
    val start = search1(nums, target, true)
    val end = search1(nums, target, false)
    ans[0] = start
    ans[1] = end
    return ans
}

fun search1(nums: Array<Int>, target: Int, findStartIndex: Boolean): Int {
    var ans = -1
    var start = 0
    var end = nums.size - 1
    while (start <= end) {
        val mid = start + (end - start) / 2

        if (target < nums[mid]) {
            end = mid - 1
        } else if (target > nums[mid]) {
            start = mid + 1
        } else {
            ans = mid
            if (findStartIndex) {
                end = mid - 1
            } else {
                start = mid + 1
            }
        }
    }
    return ans
}
